lib/Model/ConvertFunctionType.cpp

```diff
@@ -0,0 +1,664 @@
+        else if ((Index & 1) == 0 || ReturnStruct->Fields.size() <= 1
+                 || Index <= 1)
+          return std::nullopt;
+      }
+
+      MustUseTheNextOne = MustUseTheNextOne || IsCurrentRegisterUsed;
+    }
+
+    revng_assert(ReturnStruct->Size != 0 && !ReturnStruct->Fields.empty());
+    auto ReturnStructTypePath = TheBinary.recordNewType(std::move(Result));
+    revng_assert(ReturnStructTypePath.isValid());
+    return model::QualifiedType{ ReturnStructTypePath, {} };
+  }
+
+  return std::nullopt;
+}
+
+template<model::abi::Values V>
+std::optional<model::CABIFunctionType>
+convertToCABIFunctionTypeImpl(const model::RawFunctionType &Function,
+                              model::Binary &TheBinary) {
+  using CC = ::abi::CallingConventionTrait<V>;
+  static constexpr model::Architecture::Values A = getArchitecture(V);
+  if (!verify<A>(Function.Arguments, CC::GeneralPurposeArgumentRegisters))
+    return std::nullopt;
+  if (!verify<A>(Function.ReturnValues, CC::GeneralPurposeReturnValueRegisters))
+    return std::nullopt;
+
+  constexpr model::Register::Values PTCRR = CC::ReturnValueLocationRegister;
+  if (PTCRR != model::Register::Invalid)
+    revng_assert(model::Register::getArchitecture(PTCRR) == A);
+  for (auto &SavedRegister : CC::CalleeSavedRegisters)
+    revng_assert(model::Register::getArchitecture(SavedRegister) == A);
+
+  model::CABIFunctionType Result;
+  Result.ABI = V;
+
+  auto ArgumentList = convertArguments<CC>(Function.Arguments,
+                                           CC::GeneralPurposeArgumentRegisters,
+                                           TheBinary);
+  if (ArgumentList == std::nullopt)
+    return std::nullopt;
+  for (auto &Argument : *ArgumentList)
+    Result.Arguments.insert(Argument);
+
+  auto
+    ReturnValue = convertReturnValue<CC>(Function.ReturnValues,
+                                         CC::GeneralPurposeReturnValueRegisters,
+                                         CC::ReturnValueLocationRegister,
+                                         TheBinary);
+  if (ReturnValue == std::nullopt)
+    return std::nullopt;
+  Result.ReturnType = *ReturnValue;
+  return Result;
+}
+
+std::optional<model::CABIFunctionType>
+convertToCABIFunctionType(const model::RawFunctionType &Function,
+                          model::abi::Values ABI,
+                          model::Binary &TheBinary) {
+  revng_assert(ABI != model::abi::Invalid);
+  return skippingEnumSwitch<1>(ABI, [&]<model::abi::Values A>() {
+    return convertToCABIFunctionTypeImpl<A>(Function, TheBinary);
+  });
+}
+
+//
+// toRaw
+//
+
+using IndexType = decltype(model::Argument::Index);
+using RegisterList = llvm::SmallVector<model::Register::Values, 1>;
+struct DistributedArgument {
+  RegisterList Registers = {};
+  size_t Size = 0, SizeOnStack = 0;
+};
+using DistributedArguments = llvm::DenseMap<IndexType, DistributedArgument>;
+
+using ArgumentContainer = SortedVector<model::Argument>;
```

Comment:
```suggestion
using ArgumentsContainer = SortedVector<model::Argument>;
```

Reply:
Maybe I'm misunderstanding something here. But isn't `a container of arguments` -> `an argument container` the proper way to say it? As far as I understand, when nouns get chained together, all but the very last one use the singular form.

Reply:
Actually I often wonder the same question and then I think "wait, how does it work in Italian"? But maybe in English it's different!

Do you have a source to cite for what you're saying?

Reply:
It's quite hard to find a solid explanation since most of the stuff written about these chains/strings/stacks is about avoiding them rather than about formatting them. Here's the [wiki](https://en.wikipedia.org/wiki/Noun_string) entry and also a [typical "avoid those" article with a bunch of examples](https://englishcomposition.org/advanced-writing/edit-noun-strings-writing/).

Reply:
OK, as per our discussion, let's stick with `ArgumentContainer`!

Reply:
I guess this can be closed then (see the relevant discussion in the mattermost).